ydb/core/fq/libs/row_dispatcher/format_handler/filters/filters_set.cpp

```diff
@@ -8,236 +10,238 @@ namespace {
+        if (compileHandlerStatus.IsFail()) {
+            return compileHandlerStatus.GetError();
+        }
+        const auto compileHandler = compileHandlerStatus.DetachResult();
+        compileHandler->Compile();
 
-    ~TTopicFilters() {
-        Filters.clear();
+        return TStatus::Success();
     }
 
-public:
-    void FilterData(const TVector<ui64>& columnIndex, const TVector<ui64>& offsets, const TVector<const TVector<NYql::NUdf::TUnboxedValue>*>& values, ui64 numberRows) override {
-        LOG_ROW_DISPATCHER_TRACE("FilterData for " << Filters.size() << " clients, number rows: " << numberRows);
-
-        const TInstant startFilter = TInstant::Now();
-        for (const auto& [_, filterHandler] : Filters) {
-            const auto consumer = filterHandler.GetConsumer();
-            if (const auto nextOffset = consumer->GetNextMessageOffset(); !numberRows || (nextOffset && offsets[numberRows - 1] < *nextOffset)) {
-                LOG_ROW_DISPATCHER_TRACE("Ignore filtering for " << consumer->GetFilterId() << ", historical offset");
-                continue;
-            }
-            if (!filterHandler.IsStarted()) {
-                LOG_ROW_DISPATCHER_TRACE("Ignore filtering for " << consumer->GetFilterId() << ", client filter is not compiled");
-                continue;
-            }
+    TValueStatus<IProgramCompileHandler::TPtr> AddCompileProgram(TString name, IProcessedDataConsumer::TPtr consumer, IProgramHolder::TPtr programHolder, ui64 cookie) {
+        const auto clientId = consumer->GetClientId();
 
-            PushToFilter(filterHandler, offsets, columnIndex, values, numberRows);
+        const auto [inflightIter, inflightInserted] = InFlightCompilations_.emplace(cookie, std::pair{clientId, name});
+        if (!inflightInserted) {
+            return TStatus::Fail(EStatusId::INTERNAL_ERROR, "Got duplicated compilation event id");
         }
-        Stats.AddFilterLatency(TInstant::Now() - startFilter);
-    }
-
-    void OnCompileResponse(TEvRowDispatcher::TEvPurecalcCompileResponse::TPtr ev) override {
-        LOG_ROW_DISPATCHER_TRACE("Got compile response for request with id " << ev->Cookie);
 
-        const auto requestIt = InFlightCompilations.find(ev->Cookie);
-        if (requestIt == InFlightCompilations.end()) {
-            LOG_ROW_DISPATCHER_DEBUG("Compile response ignored for id " << ev->Cookie);
-            return;
+        auto compileHandler = CreateProgramCompileHandler(name, std::move(consumer), std::move(programHolder), cookie, Config_.CompileServiceId, Owner_, Counters_.Counters);
+        auto& compileHandlers = CompileHandlers_[clientId];
+        auto [iter, inserted] = compileHandlers.emplace(name, std::move(compileHandler));
+        if (!inserted) {
+            InFlightCompilations_.erase(inflightIter);
+            return TStatus::Fail(EStatusId::INTERNAL_ERROR, TStringBuilder() << "Failed to compile new program, program with client id " << clientId << " already exists");
```

Comment:
Тут кажется можно добавить ещё name в ошибку

```suggestion
            return TStatus::Fail(EStatusId::INTERNAL_ERROR, TStringBuilder() << "Failed to compile new program, program with client id " << clientId << " and name " << name << " already exists");
```